File: PruebaPicasso/app/src/main/java/es/iesnervion/avazquez/pruebapicasso/viewModel/SharedViewModel.java
package es.iesnervion.avazquez.pruebapicasso.viewModel;

import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.io.IOException;
import java.util.ArrayList;

import es.iesnervion.avazquez.pruebapicasso.R;
import es.iesnervion.avazquez.pruebapicasso.RespuestaAPI;
import es.iesnervion.avazquez.pruebapicasso.interfaces.APIInterface;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class SharedViewModel extends ViewModel {
    private MutableLiveData<ArrayList<String>> listaURLs;
    private MutableLiveData<String> razaElegida;
    private MutableLiveData<Boolean> isCargada;

    //Constructor
    public SharedViewModel() {
        this.listaURLs = new MutableLiveData<>();
        this.razaElegida = new MutableLiveData<>();
        this.isCargada = new MutableLiveData<>();
    }

    //getters y setters
    public LiveData<String> getRazaElegida(){
        if(this.razaElegida == null){
            this.razaElegida = new MutableLiveData<>();
        }
        return this.razaElegida;
    }

    public LiveData<Boolean> isCargada(){
        if(this.isCargada == null){
            this.isCargada = new MutableLiveData<>();
        }
        return this.isCargada;
    }

    public void setIsCargada(boolean isCargada){
        //this.isCargada.setValue(isCargada);
        this.isCargada.postValue(isCargada);
    }

    public void setRazaElegida(String razaElegida){
        this.razaElegida.setValue(razaElegida);
        cargarListaURLs();

    }

    public LiveData<ArrayList<String>> getListaURLs(){
        if(this.listaURLs == null ){
            this.listaURLs = new MutableLiveData<>();
        //no puedes poner el cargarListaURLs aqui porque listaURLs no sera null. Si acaso
            //tendrias que poner si listaURLs.value es null
        }
//        cargarListaURLs();
        return this.listaURLs;
    }


    public void setListaURLs(ArrayList<String> listaURLs) {
        this.listaURLs.setValue(listaURLs);
    }

    private void cargarListaURLs() {


        if(this.getRazaElegida().getValue() != null){
            Retrofit retrofit;
            retrofit = new Retrofit.Builder()
                    .baseUrl("https://dog.ceo/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();

            APIInterface apiInterface = retrofit.create(APIInterface.class);

            Call<RespuestaAPI> call = apiInterface
                    .getFotosDeRaza(this.getRazaElegida().getValue());

//            TODO esto es asíncrono y entonces el view model recibe un null porque no se espera

            //https://stackoverflow.com/questions/47553632/wait-for-retrofit-asynchronous-call-response-on-an-adapter
            call.enqueue(new Callback<RespuestaAPI>() {
                @Override
                public void onResponse(Call<RespuestaAPI> call, Response<RespuestaAPI> response) {

                    if(response.isSuccessful()){
                        setListaURLs((ArrayList<String>) response.body().getLista());
                        //TODO haz aqui el cambio de fragment
                        setIsCargada(true);

                    }else{
                        ArrayList img = new ArrayList();
                        img.add("https://love2dev.com/img/error-message-laptop-1920x1297.jpg");

                        setListaURLs(img);
                    }

                }

                @Override
                public void onFailure(Call<RespuestaAPI> call, Throwable t) {
                    t.getStackTrace();
                    ArrayList img = new ArrayList();
                    img.add("https://love2dev.com/img/error-message-laptop-1920x1297.jpg");

                    setListaURLs(img);
                }
            });

        }
    }


}

File: EjemploKotlinRetrofit2/app/src/main/java/es/iesnervion/avazquez/ejemplokotlinretrofit2/viewModel/SharedViewModel.kt
package es.iesnervion.avazquez.ejemplokotlinretrofit2.viewModel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import es.iesnervion.avazquez.ejemplokotlinretrofit2.entities.Perrete
import es.iesnervion.avazquez.ejemplokotlinretrofit2.interfaces.PerroInterface
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class SharedViewModel : ViewModel() {
    private var fotos : MutableLiveData<ArrayList<String>>
    private var razaElegida : MutableLiveData<String>

    init{
        this.razaElegida = MutableLiveData()
        this.fotos = MutableLiveData()
    }

    fun setRazaElegida(razaElegida :String){
        this.razaElegida.value = razaElegida
        cargarFotos()
    }

    fun getRazaElegida() : MutableLiveData<String>{
        return this.razaElegida
    }


    fun getURLsFotos(): ArrayList<String> {

//        if(this.fotos.value != null && this.fotos.value!!.size > 0){
//            this.fotos.value = ArrayList()
//        }

        this.fotos.value = arrayListOf( "https://images.dog.ceo/breeds/hound-afghan/n02088094_1003.jpg",
            "https://images.dog.ceo/breeds/hound-afghan/n02088094_1007.jpg",
            "https://images.dog.ceo/breeds/hound-afghan/n02088094_1023.jpg",
            "https://images.dog.ceo/breeds/hound-afghan/n02088094_10263.jpg",
            "https://images.dog.ceo/breeds/hound-afghan/n02088094_10715.jpg",
            "https://images.dog.ceo/breeds/hound-afghan/n02088094_10822.jpg")
        return this.fotos.value!!
    }

    fun getMutableLiveDataFotos() : MutableLiveData<ArrayList<String>>{
        return this.fotos
    }

    private fun cargarFotos() {
        /*Aqui hago lo de la API*/
        var retrofit : Retrofit =
            Retrofit.Builder()
                .baseUrl("https://dog.ceo/")
                .addConverterFactory(GsonConverterFactory.create())
                .build()

        var perroInterface = retrofit.create(PerroInterface::class.java)

        var call: Call<Perrete> = perroInterface.getFotosDeRaza(razaElegida.value!!)

        call.enqueue(object : Callback<Perrete> {
            override fun onFailure(call: Call<Perrete>, t: Throwable) {
                t.printStackTrace()
            }

            override fun onResponse(call: Call<Perrete>, response: Response<Perrete>) {
                fotos.value = ArrayList(response.body()?.fotos!!)
            }
        })






    }



}